hal/src/rtc/modes.rs

```diff
@@ -0,0 +1,337 @@
+//! - Handles ATSAMD chip variations.
+//!
+//! The idea is that various higher-level users of these abstractions will not
+//! have to handle these low-level aspects of using the RTC. However, this
+//! module does not present a safe interface. For example, many of the methods
+//! in [`RtcMode`] assume that the RTC has already been put into the correct
+//! mode (using [`RtcMode::set_mode`]), but without enforcing this in any way.
+
+// As explained in the [datasheets](https://onlinedocs.microchip.com/oxy/GUID-F5813793-E016-46F5-A9E2-718D8BCED496-en-US-14/GUID-ABE2D37F-8125-4279-9955-BC3900046CFF.html),
+// reading a read-synced register may result in
+// an old value, which we try to avoid by ensuring that SYNCBUSY is clear
+// before reading. A write to a write-synced register will be discarded if
+// syncing is happening during the write. As such, we also ensure that SYNCBUSY
+// is clear before writing to a synced register. Throughout the crate, every
+// register access should be prefaced by a `SYNC` comment indicating the
+// required synchronization. The presence of this comment signals that this
+// access was checked in the datasheet and accounted for.
+
+use crate::pac;
+use atsamd_hal_macros::hal_macro_helper;
+use pac::Rtc;
+
+/// Type-level enum for RTC interrupts.
+pub trait RtcInterrupt {
+    /// Enable this interrupt.
+    fn enable(rtc: &Rtc);
+    /// Returns whether the interrupt has been triggered.
+    fn check_flag(rtc: &Rtc) -> bool;
+    /// Clears the interrupt flag so the ISR will not be called again
+    /// immediately.
+    fn clear_flag(rtc: &Rtc);
+}
+
+/// Macro to easily declare an RTC interrupt.
+macro_rules! create_rtc_interrupt {
+    ($mode:ident, $name:ident, $bit:ident) => {
+        /// Type-level variant for the $name interrupt in $mode.
+        pub enum $name {}
+        impl RtcInterrupt for $name {
+            #[inline]
+            fn enable(rtc: &Rtc) {
+                // SYNC: None
+                rtc.$mode().intenset().write(|w| w.$bit().set_bit());
+            }
+
+            #[inline]
+            fn check_flag(rtc: &Rtc) -> bool {
+                // SYNC: None
+                rtc.$mode().intflag().read().$bit().bit_is_set()
+            }
+
+            #[inline]
+            fn clear_flag(rtc: &Rtc) {
+                // SYNC: None
+                rtc.$mode().intflag().write(|w| w.$bit().set_bit());
+            }
+        }
+    };
+}
+
+/// An abstraction of an RTC in a particular mode that provides low-level
+/// access and handles all register syncing issues using only associated
+/// functions.
+#[hal_macro_helper]
+pub trait RtcMode {
+    /// The type of the COUNT register.
+    type Count: Copy + PartialEq + Eq;
+
+    /// Sets this mode in the CTRL register.
+    unsafe fn set_mode(rtc: &Rtc);
+    /// Sets a compare value.
+    unsafe fn set_compare(rtc: &Rtc, number: usize, value: Self::Count);
```

Comment:
Here, I would suggest adding a `# Safety` section to the doc comments, explaining why these functions are unsafe and what are the invariants that must be upheld.